driver-core/src/main/com/mongodb/internal/connection/ConcurrentPool.java

```diff
@@ -501,4 +537,45 @@ boolean closed() {
     static String sizeToString(final int size) {
         return size == INFINITE_SIZE ? "infinite" : Integer.toString(size);
     }
+
+    static void lockInterruptibly(final Lock lock) throws MongoInterruptedException {
+        try {
+            lock.lockInterruptibly();
+        } catch (InterruptedException e) {
+            throw new MongoInterruptedException(null, e);
+        }
+    }
+
+    private static void lockInterruptiblyUnfair(final ReentrantReadWriteLock.WriteLock lock) throws MongoInterruptedException {
+        throwIfInterrupted();
+        // `WriteLock.tryLock` is unfair
+        if (!lock.tryLock()) {
+            try {
+                lock.lockInterruptibly();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new MongoInterruptedException(null, new InterruptedException());
+            }
+        }
+    }
+
+    static void lockUnfair(final ReentrantLock lock) {
+        // `ReentrantLock.tryLock` is unfair
+        if (!lock.tryLock()) {
+            lock.lock();
+        }
+    }
+
+    private static void lockUnfair(final ReentrantReadWriteLock.WriteLock lock) {
+        // `WriteLock.tryLock` is unfair
+        if (!lock.tryLock()) {
+            lock.lock();
+        }
+    }
```

Comment:
See the discussion [here](https://github.com/mongodb/mongo-java-driver/pull/857#discussion_r791658469) about the reason behind introducing two methods as opposed to having one that accepts `Lock`.